app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt

```diff
@@ -137,130 +159,92 @@ class LoginActivity : BaseActivity(), Injectable {
             return
         }
 
-        val callback = object : Callback<AppCredentials> {
-            override fun onResponse(
-                call: Call<AppCredentials>,
-                response: Response<AppCredentials>
-            ) {
-                if (!response.isSuccessful) {
-                    binding.loginButton.isEnabled = true
-                    binding.domainTextInputLayout.error = getString(R.string.error_failed_app_registration)
-                    setLoading(false)
-                    Log.e(TAG, "App authentication failed. " + response.message())
-                    return
-                }
-                val credentials = response.body()
-                val clientId = credentials!!.clientId
-                val clientSecret = credentials.clientSecret
-
-                preferences.edit()
-                    .putString("domain", domain)
-                    .putString("clientId", clientId)
-                    .putString("clientSecret", clientSecret)
-                    .apply()
-
-                redirectUserToAuthorizeAndLogin(domain, clientId)
-            }
+        setLoading(true)
 
-            override fun onFailure(call: Call<AppCredentials>, t: Throwable) {
+        lifecycleScope.launch {
+            val credentials: AppCredentials = try {
+                mastodonApi.authenticateApp(
+                    domain, getString(R.string.app_name), oauthRedirectUri.toString(),
```

Comment:
`oauthRedirectUri` already is a String
(also 3 more times below)

Reply:
Oh yeah I made it a `HttpUrl` but it's not actually a "http" URL so I rolled it back.